hibernate-core/src/main/java/org/hibernate/query/hql/internal/QualifiedJoinPathConsumer.java

```diff
@@ -178,8 +178,12 @@ private static SqmFrom<?, Object> createJoin(
 			boolean isTerminal,
 			boolean allowReuse,
 			SqmCreationState creationState) {
+		final SqmPathSource<?> referencedPathSource = lhs.getReferencedPathSource();
+		final SqmPathSource<?> pathSource = referencedPathSource.isGeneric() ?
```

Comment:
I'm a bit confused, shouldn't `lhs.getResolvedModel()` always produce the desired outcome?

Reply:
Not for treated joins, since `getResolvedModel()` for non-generic paths returns `getModel()` we would get the non-treated association attribute as a path source, which would report a wrong associated entity type (first one in alphabetical order).

Reply:
We could override this behavior in the `SqmTreatedPath` interface with a default `getResolvedModel()` implementation that returns the treat target entity type, but I wasn't sure if that is correct from the JPA-model side.

Reply:
Ok, makes sense. Thanks

Reply:
I've added a comment to clarify why we need to use the referenced path source, let's see if a `getResolvedNodeType()` or similar could be useful in other places in the future.